File: backend/app/domain/services/snapshot.py
import uuid
from datetime import datetime, timezone
from typing import Any

from app.domain.entities.market import AssetSnapshot, PriceHistory
from app.domain.services.base import BaseService
from app.infrastructure.repositories.asset_snapshot import AssetSnapshotRepository
from app.infrastructure.repositories.market import MarketRepository


class SnapshotService(BaseService):
    def __init__(self, snapshot_repo: AssetSnapshotRepository, market_repo: MarketRepository):
        self.snapshot_repo = snapshot_repo
        self.market_repo = market_repo

    def build_snapshot(self, asset_id: uuid.UUID, indicators: dict[str, Any]) -> dict[str, Any] | None:
        """Persists an AssetSnapshot + PriceHistory point from the latest quote and
        pre-computed technical indicators. Returns the Redis cache payload, or None
        if there is no quote to snapshot yet."""
        quote = self.market_repo.get_quote_by_asset_id(asset_id)
        price = float(quote.price) if quote and quote.price is not None else None
        volume = float(quote.volume) if quote and quote.volume is not None else None
        symbol = quote.symbol if quote else None

        rsi_val = indicators.get("rsi")
        momentum_val = rsi_val / 100.0 if rsi_val is not None else None
        volatility_val = indicators.get("volatility")
        sentiment_val = indicators.get("sentiment")

        snapshot = AssetSnapshot(
            asset_id=asset_id,
            price=price,
            market_cap=None,
            pe_ratio=None,
            rsi=rsi_val,
            momentum_score=momentum_val,
            volatility_score=volatility_val,
            sentiment_score=sentiment_val,
            payload=indicators or {},
            updated_at=datetime.now(timezone.utc)
        )
        updated_snapshot = self.snapshot_repo.upsert(snapshot)

        if price is not None and symbol is not None:
            self.market_repo.add_price_history(PriceHistory(
                id=uuid.uuid4(),
                asset_id=asset_id,
                symbol=symbol,
                price=price,
                volume=volume,
                timestamp=datetime.now(timezone.utc)
            ))

        self.snapshot_repo.session.commit()

        return {
            "asset_id": str(updated_snapshot.asset_id),
            "price": float(updated_snapshot.price) if updated_snapshot.price is not None else None,
            "market_cap": float(updated_snapshot.market_cap) if updated_snapshot.market_cap is not None else None,
            "pe_ratio": float(updated_snapshot.pe_ratio) if updated_snapshot.pe_ratio is not None else None,
            "rsi": float(updated_snapshot.rsi) if updated_snapshot.rsi is not None else None,
            "momentum_score": float(updated_snapshot.momentum_score) if updated_snapshot.momentum_score is not None else None,
            "volatility_score": float(updated_snapshot.volatility_score) if updated_snapshot.volatility_score is not None else None,
            "sentiment_score": float(updated_snapshot.sentiment_score) if updated_snapshot.sentiment_score is not None else None,
            "payload": updated_snapshot.payload,
            "updated_at": updated_snapshot.updated_at.isoformat() if updated_snapshot.updated_at else None
        }

File: backend/src/lib/evaluation/snapshot.ts
import { v4 as uuidv4 } from "uuid";
import { prisma } from "../../prisma";
import { Prisma } from "../../generated/prisma";
import type { TechnicalIndicators } from "../marketProviders/yahoo";

export interface AssetSnapshotResult {
  asset_id: string;
  price: number | null;
  market_cap: number | null;
  pe_ratio: number | null;
  rsi: number | null;
  momentum_score: number | null;
  volatility_score: number | null;
  sentiment_score: number | null;
  payload: unknown;
  updated_at: string | null;
}

/** Port of SnapshotService.build_snapshot. Persists an AssetSnapshot +
 * PriceHistory point from the latest quote and pre-computed technical
 * indicators, and returns the Redis cache payload — unconditionally: despite
 * Python's own docstring claiming "or None if there is no quote to snapshot
 * yet," the actual Python code never returns None (there's no early-return
 * branch), it just upserts a snapshot with a null price. Matching the real
 * code, not the stale docstring.
 *
 * PriceHistory.id here is a random uuid4, not the uuid5(symbol+date) dedup
 * convention used by ingestQuote/seedPriceHistory/refreshMutualFundNavs —
 * matches Python's asset_snapshot.py exactly (a real, pre-existing
 * inconsistency in the Python source, not something to "fix" while porting:
 * every process_asset_snapshot run adds a fresh row, same day or not). */
export async function buildAssetSnapshot(assetId: string, indicators: Partial<TechnicalIndicators>): Promise<AssetSnapshotResult> {
  const quote = await prisma.latestQuote.findFirst({ where: { assetId } });
  const price = quote?.price != null ? Number(quote.price) : null;
  const volume = quote?.volume != null ? Number(quote.volume) : null;
  const symbol = quote?.symbol ?? null;

  const rsiVal = indicators.rsi ?? null;
  const momentumVal = rsiVal !== null ? rsiVal / 100.0 : null;
  const volatilityVal = indicators.volatility ?? null;
  const sentimentVal = indicators.sentiment ?? null;

  const now = new Date();
  const updated = await prisma.assetSnapshot.upsert({
    where: { assetId },
    create: {
      assetId,
      price,
      marketCap: null,
      peRatio: null,
      rsi: rsiVal,
      momentumScore: momentumVal,
      volatilityScore: volatilityVal,
      sentimentScore: sentimentVal,
      payload: (indicators as Prisma.InputJsonValue) ?? {},
      createdAt: now,
      updatedAt: now,
    },
    update: {
      price,
      rsi: rsiVal,
      momentumScore: momentumVal,
      volatilityScore: volatilityVal,
      sentimentScore: sentimentVal,
      payload: (indicators as Prisma.InputJsonValue) ?? {},
      updatedAt: now,
    },
  });

  if (price !== null && symbol !== null) {
    await prisma.priceHistory.create({
      data: { id: uuidv4(), assetId, symbol, price, volume, timestamp: now },
    });
  }

  return {
    asset_id: updated.assetId,
    price: updated.price != null ? Number(updated.price) : null,
    market_cap: updated.marketCap != null ? Number(updated.marketCap) : null,
    pe_ratio: updated.peRatio != null ? Number(updated.peRatio) : null,
    rsi: updated.rsi != null ? Number(updated.rsi) : null,
    momentum_score: updated.momentumScore != null ? Number(updated.momentumScore) : null,
    volatility_score: updated.volatilityScore != null ? Number(updated.volatilityScore) : null,
    sentiment_score: updated.sentimentScore != null ? Number(updated.sentimentScore) : null,
    payload: updated.payload,
    updated_at: updated.updatedAt.toISOString(),
  };
}
